In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('netflix.csv')

# Understanding the Data

## Gist of dataset

In [3]:
df.head(10)

show_id     type                             title  \
0      s1    Movie              Dick Johnson Is Dead   
1      s2  TV Show                     Blood & Water   
2      s3  TV Show                         Ganglands   
3      s4  TV Show             Jailbirds New Orleans   
4      s5  TV Show                      Kota Factory   
5      s6  TV Show                     Midnight Mass   
6      s7    Movie  My Little Pony: A New Generation   
7      s8    Movie                           Sankofa   
8      s9  TV Show     The Great British Baking Show   
9     s10    Movie                      The Starling   

                        director  \
0                Kirsten Johnson   
1                            NaN   
2                Julien Leclercq   
3                            NaN   
4                            NaN   
5                  Mike Flanagan   
6  Robert Cullen, José Luis Ucha   
7                   Haile Gerima   
8                Andy Devonshire   
9                 Theodore Melfi   

                                                cast  \
0                                                NaN   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   

                                             country          date_added  \
0                                      United States  September 25, 2021   
1                                       South Africa  September 24, 2021   
2                                                NaN  September 24, 2021   
3                                                NaN  September 24, 2021   
4                                              India  September 24, 2021   
5                                                NaN  September 24, 2021   
6                                                NaN  September 24, 2021   
7  United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
8                                     United Kingdom  September 24, 2021   
9                                      United States  September 24, 2021   

   release_year rating   duration  \
0          2020  PG-13     90 min   
1          2021  TV-MA  2 Seasons   
2          2021  TV-MA   1 Season   
3          2021  TV-MA   1 Season   
4          2021  TV-MA  2 Seasons   
5          2021  TV-MA   1 Season   
6          2021     PG     91 min   
7          1993  TV-MA    125 min   
8          2021  TV-14  9 Seasons   
9          2021  PG-13    104 min   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   
5                 TV Dramas, TV Horror, TV Mysteries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
8                       British TV Shows, Reality TV   
9                                   Comedies, Dramas   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...  
5  The arrival of a charismatic young priest brin...  
6  Equestria's divided. But a bright-eyed hero be...  
7  On a photo shoot in Ghana, an American model s...  
8  A talented 

In [4]:
df.shape

(8807, 12)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


## Challenges with the data

### 1. Missing Values

In [6]:
na_variables = [ var for var in df.columns if df[var].isnull().mean() > 0 ]
na_variables

['director', 'cast', 'country', 'date_added', 'rating', 'duration']

In [7]:
df[na_variables].isnull().mean()*100

director      29.908028
cast           9.367549
country        9.435676
date_added     0.113546
rating         0.045418
duration       0.034064
dtype: float64

In [8]:
df.isnull().sum()/len(df)*100

show_id          0.000000
type             0.000000
title            0.000000
director        29.908028
cast             9.367549
country          9.435676
date_added       0.113546
release_year     0.000000
rating           0.045418
duration         0.034064
listed_in        0.000000
description      0.000000
dtype: float64

#### We can see that the columns which are significantly impacted by null values are director, cast and country with director being impacted at a very high level of almost 30%.
#### Apart from the above 3 columns, date_added, rating and duration also have missing values but the volume is not significant. Hence we can consider removing those data/rows from the dataset.

#### a) Analysing and removing missing values for date_added, rating and duration

In [9]:
print('Percent of records which are null for the columns date_added, rating and duration: ', df.iloc[np.where((pd.isna(df['date_added'])==True) | (pd.isna(df['rating'])==True) | (pd.isna(df['duration'])==True))].shape[0]/len(df)*100)

Percent of records which are null for the columns date_added, rating and duration:  0.19302827296468716


In [10]:
df.dropna(subset = ['date_added', 'rating', 'duration'], inplace = True)

In [11]:
df.shape

(8790, 12)

#### 17 records removed out of total 8807 making the new total record count of 8790

#### b) Analysing missing "director" values

In [12]:
df.iloc[np.where(pd.isna(df['director'])==True)].head()

show_id     type                                title director  \
1       s2  TV Show                        Blood & Water      NaN   
3       s4  TV Show                Jailbirds New Orleans      NaN   
4       s5  TV Show                         Kota Factory      NaN   
10     s11  TV Show  Vendetta: Truth, Lies and The Mafia      NaN   
14     s15  TV Show      Crime Stories: India Detectives      NaN   

                                                 cast       country  \
1   Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...  South Africa   
3                                                 NaN           NaN   
4   Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...         India   
10                                                NaN           NaN   
14                                                NaN           NaN   

            date_added  release_year rating   duration  \
1   September 24, 2021          2021  TV-MA  2 Seasons   
3   September 24, 2021          2021  TV-MA   1 Season   
4   September 24, 2021          2021  TV-MA  2 Seasons   
10  September 24, 2021          2021  TV-MA   1 Season   
14  September 22, 2021          2021  TV-MA   1 Season   

                                            listed_in  \
1     International TV Shows, TV Dramas, TV Mysteries   
3                              Docuseries, Reality TV   
4   International TV Shows, Romantic TV Shows, TV ...   
10  Crime TV Shows, Docuseries, International TV S...   
14       British TV Shows, Crime TV Shows, Docuseries   

                                          description  
1   After crossing paths at a party, a Cape Town t...  
3   Feuds, flirtations and toilet talk go down amo...  
4   In a city of coaching centers known to train I...  
10  Sicily boasts a bold "Anti-Mafia" coalition. B...  
14  Cameras following Bengaluru police on the job ...

In [13]:
df.iloc[np.where(pd.isna(df['director'])==True)].shape

(2621, 12)

In [14]:
print('Percentage of significant null values in director column:', round((df.isnull().sum()/len(df)*100)['director'].astype(float),2))

Percentage of significant null values in director column: 29.82


#### c) Analysing missing "cast" values

In [15]:
df.iloc[np.where(pd.isna(df['cast'])==True)].head()

show_id     type                                              title  \
0       s1    Movie                               Dick Johnson Is Dead   
3       s4  TV Show                              Jailbirds New Orleans   
10     s11  TV Show                Vendetta: Truth, Lies and The Mafia   
14     s15  TV Show                    Crime Stories: India Detectives   
16     s17    Movie  Europe's Most Dangerous Man: Otto Skorzeny in ...   

                                         director cast        country  \
0                                 Kirsten Johnson  NaN  United States   
3                                             NaN  NaN            NaN   
10                                            NaN  NaN            NaN   
14                                            NaN  NaN            NaN   
16  Pedro de Echave García, Pablo Azorín Williams  NaN            NaN   

            date_added  release_year rating  duration  \
0   September 25, 2021          2020  PG-13    90 min   
3   September 24, 2021          2021  TV-MA  1 Season   
10  September 24, 2021          2021  TV-MA  1 Season   
14  September 22, 2021          2021  TV-MA  1 Season   
16  September 22, 2021          2020  TV-MA    67 min   

                                            listed_in  \
0                                       Documentaries   
3                              Docuseries, Reality TV   
10  Crime TV Shows, Docuseries, International TV S...   
14       British TV Shows, Crime TV Shows, Docuseries   
16                Documentaries, International Movies   

                                          description  
0   As her father nears the end of his life, filmm...  
3   Feuds, flirtations and toilet talk go down amo...  
10  Sicily boasts a bold "Anti-Mafia" coalition. B...  
14  Cameras following Bengaluru police on the job ...  
16  Declassified documents reveal the post-WWII li...

In [16]:
df.iloc[np.where(pd.isna(df['cast'])==True)].shape

(825, 12)

#### We can ignore the case where cast is NaN for listed_in values Documentaries as usually, they may or may not have cast

In [17]:
df.iloc[np.where((df['listed_in']!='Documentaries') & (pd.isna(df['cast'])==True))].shape

(642, 12)

In [18]:
df.iloc[np.where((df['listed_in']!='Documentaries') & (pd.isna(df['cast'])==True))].head()

show_id     type                                              title  \
3       s4  TV Show                              Jailbirds New Orleans   
10     s11  TV Show                Vendetta: Truth, Lies and The Mafia   
14     s15  TV Show                    Crime Stories: India Detectives   
16     s17    Movie  Europe's Most Dangerous Man: Otto Skorzeny in ...   
20     s21  TV Show    Monsters Inside: The 24 Faces of Billy Milligan   

                                         director cast country  \
3                                             NaN  NaN     NaN   
10                                            NaN  NaN     NaN   
14                                            NaN  NaN     NaN   
16  Pedro de Echave García, Pablo Azorín Williams  NaN     NaN   
20                                Olivier Megaton  NaN     NaN   

            date_added  release_year rating  duration  \
3   September 24, 2021          2021  TV-MA  1 Season   
10  September 24, 2021          2021  TV-MA  1 Season   
14  September 22, 2021          2021  TV-MA  1 Season   
16  September 22, 2021          2020  TV-MA    67 min   
20  September 22, 2021          2021  TV-14  1 Season   

                                            listed_in  \
3                              Docuseries, Reality TV   
10  Crime TV Shows, Docuseries, International TV S...   
14       British TV Shows, Crime TV Shows, Docuseries   
16                Documentaries, International Movies   
20  Crime TV Shows, Docuseries, International TV S...   

                                          description  
3   Feuds, flirtations and toilet talk go down amo...  
10  Sicily boasts a bold "Anti-Mafia" coalition. B...  
14  Cameras following Bengaluru police on the job ...  
16  Declassified documents reveal the post-WWII li...  
20  In the late 1970s, an accused serial rapist cl...

In [19]:
print('Percentage of significant null values in cast column:', round(df.iloc[np.where((df['listed_in']!='Documentaries') & (pd.isna(df['cast'])==True))].shape[0]/len(df)*100, 2))

Percentage of significant null values in cast column: 7.3


#### d) Analysing missing "country" values

In [20]:
df.iloc[np.where(pd.isna(df['country'])==True)].shape

(829, 12)

In [21]:
df.iloc[np.where(pd.isna(df['country'])==True)].head()

show_id     type                                title  \
2       s3  TV Show                            Ganglands   
3       s4  TV Show                Jailbirds New Orleans   
5       s6  TV Show                        Midnight Mass   
6       s7    Movie     My Little Pony: A New Generation   
10     s11  TV Show  Vendetta: Truth, Lies and The Mafia   

                         director  \
2                 Julien Leclercq   
3                             NaN   
5                   Mike Flanagan   
6   Robert Cullen, José Luis Ucha   
10                            NaN   

                                                 cast country  \
2   Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...     NaN   
3                                                 NaN     NaN   
5   Kate Siegel, Zach Gilford, Hamish Linklater, H...     NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...     NaN   
10                                                NaN     NaN   

            date_added  release_year rating  duration  \
2   September 24, 2021          2021  TV-MA  1 Season   
3   September 24, 2021          2021  TV-MA  1 Season   
5   September 24, 2021          2021  TV-MA  1 Season   
6   September 24, 2021          2021     PG    91 min   
10  September 24, 2021          2021  TV-MA  1 Season   

                                            listed_in  \
2   Crime TV Shows, International TV Shows, TV Act...   
3                              Docuseries, Reality TV   
5                  TV Dramas, TV Horror, TV Mysteries   
6                            Children & Family Movies   
10  Crime TV Shows, Docuseries, International TV S...   

                                          description  
2   To protect his family from a powerful drug lor...  
3   Feuds, flirtations and toilet talk go down amo...  
5   The arrival of a charismatic young priest brin...  
6   Equestria's divided. But a bright-eyed hero be...  
10  Sicily boasts a bold "Anti-Mafia" coalition. B...

In [22]:
print('Percentage of significant null values in country column:', round(df.iloc[np.where(pd.isna(df['country'])==True)].shape[0]/len(df)*100, 2))

Percentage of significant null values in country column: 9.43


In [23]:
df.shape

(8790, 12)

### 2. Nested data issue

#### It's been identified that we have nested data issue with the columns director, cast, country and listed_in.

#### a) Unnesting Director

In [24]:
df = df.assign(director = df['director'].str.split(', ')).explode('director')

In [25]:
df = df.reset_index()

In [26]:
df.drop('index', axis=1, inplace=True)

In [27]:
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
9590   s8803    Movie                 Zodiac    David Fincher   
9591   s8804  TV Show            Zombie Dumb              NaN   
9592   s8805    Movie             Zombieland  Ruben Fleischer   
9593   s8806    Movie                   Zoom     Peter Hewitt   
9594   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
9590  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
9591                                                NaN            NaN   
9592  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
9593  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
9594  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
9590   November 20, 2019          2007      R    158 min   
9591        July 1, 2019          2018  TV-Y7  2 Seasons   
9592    November 1, 2019          2009      R     88 min   
9593    January 11, 2020          2006     PG     88 min   
9594       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
9590                     Cult Movies, Dramas, Thrillers   
9591             Kids' TV, Korean TV Shows, TV Comedies   
9592                            Comedies, Horror Movies   
9593                 Children & Family Movies, Comedies   
9594     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
9590  A political cartoonist, a crime reporter and a...  
9591  While living alone in a spooky town, a young g...  
9592  Looking to survive in a world taken over by zo...  
9593  Dragged from civilian life, a former superhero...  
9594  A scrappy but poor boy worms his way into a ty...  

[9595 rows x 12 columns]

#### Column director has been unnested

#### b) Unnesting cast

In [28]:
df = df.assign(cast = df['cast'].str.split(', ')).explode('cast')

In [29]:
df = df.reset_index()

In [30]:
df.drop('index', axis=1, inplace=True)

In [31]:
df

show_id     type                 title         director  \
0          s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1          s2  TV Show         Blood & Water              NaN   
2          s2  TV Show         Blood & Water              NaN   
3          s2  TV Show         Blood & Water              NaN   
4          s2  TV Show         Blood & Water              NaN   
...       ...      ...                   ...              ...   
70697   s8807    Movie                Zubaan      Mozez Singh   
70698   s8807    Movie                Zubaan      Mozez Singh   
70699   s8807    Movie                Zubaan      Mozez Singh   
70700   s8807    Movie                Zubaan      Mozez Singh   
70701   s8807    Movie                Zubaan      Mozez Singh   

                        cast        country          date_added  release_year  \
0                        NaN  United States  September 25, 2021          2020   
1                 Ama Qamata   South Africa  September 24, 2021          2021   
2                Khosi Ngema   South Africa  September 24, 2021          2021   
3              Gail Mabalane   South Africa  September 24, 2021          2021   
4             Thabang Molaba   South Africa  September 24, 2021          2021   
...                      ...            ...                 ...           ...   
70697       Manish Chaudhary          India       March 2, 2019          2015   
70698           Meghna Malik          India       March 2, 2019          2015   
70699          Malkeet Rauni          India       March 2, 2019          2015   
70700         Anita Shabdish          India       March 2, 2019          2015   
70701  Chittaranjan Tripathy          India       March 2, 2019          2015   

      rating   duration                                        listed_in  \
0      PG-13     90 min                                    Documentaries   
1      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
2      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
3      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
4      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
...      ...        ...                                              ...   
70697  TV-14    111 min   Dramas, International Movies, Music & Musicals   
70698  TV-14    111 min   Dramas, International Movies, Music & Musicals   
70699  TV-14    111 min   Dramas, International Movies, Music & Musicals   
70700  TV-14    111 min   Dramas, International Movies, Music & Musicals   
70701  TV-14    111 min   Dramas, International Movies, Music & Musicals   

                                             description  
0      As her father nears the end of his life, filmm...  
1      After crossing paths at a party, a Cape Town t...  
2      After crossing paths at a party, a Cape Town t...  
3      After crossing paths at a party, a Cape Town t...  
4      After crossing paths at a party, a Cape Town t...  
...                                                  ...  
70697  A scrappy but poor boy worms his way into a ty...  
70698  A scrappy but poor boy worms his way into a ty...  
70699  A scrappy but poor boy worms his way into a ty...  
70700  A scrappy but poor boy worms his way into a ty...  
70701  A scrappy but poor boy worms his way into a ty...  

[70702 rows x 12 columns]

#### Column cast has been unnested

#### c) Unnesting country

In [32]:
df = df.assign(country = df['country'].str.split(', ')).explode('country')

In [33]:
df = df.reset_index()

In [34]:
df.drop('index', axis=1, inplace=True)

In [35]:
df

show_id     type                 title         director  \
0          s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1          s2  TV Show         Blood & Water              NaN   
2          s2  TV Show         Blood & Water              NaN   
3          s2  TV Show         Blood & Water              NaN   
4          s2  TV Show         Blood & Water              NaN   
...       ...      ...                   ...              ...   
89267   s8807    Movie                Zubaan      Mozez Singh   
89268   s8807    Movie                Zubaan      Mozez Singh   
89269   s8807    Movie                Zubaan      Mozez Singh   
89270   s8807    Movie                Zubaan      Mozez Singh   
89271   s8807    Movie                Zubaan      Mozez Singh   

                        cast        country          date_added  release_year  \
0                        NaN  United States  September 25, 2021          2020   
1                 Ama Qamata   South Africa  September 24, 2021          2021   
2                Khosi Ngema   South Africa  September 24, 2021          2021   
3              Gail Mabalane   South Africa  September 24, 2021          2021   
4             Thabang Molaba   South Africa  September 24, 2021          2021   
...                      ...            ...                 ...           ...   
89267       Manish Chaudhary          India       March 2, 2019          2015   
89268           Meghna Malik          India       March 2, 2019          2015   
89269          Malkeet Rauni          India       March 2, 2019          2015   
89270         Anita Shabdish          India       March 2, 2019          2015   
89271  Chittaranjan Tripathy          India       March 2, 2019          2015   

      rating   duration                                        listed_in  \
0      PG-13     90 min                                    Documentaries   
1      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
2      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
3      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
4      TV-MA  2 Seasons  International TV Shows, TV Dramas, TV Mysteries   
...      ...        ...                                              ...   
89267  TV-14    111 min   Dramas, International Movies, Music & Musicals   
89268  TV-14    111 min   Dramas, International Movies, Music & Musicals   
89269  TV-14    111 min   Dramas, International Movies, Music & Musicals   
89270  TV-14    111 min   Dramas, International Movies, Music & Musicals   
89271  TV-14    111 min   Dramas, International Movies, Music & Musicals   

                                             description  
0      As her father nears the end of his life, filmm...  
1      After crossing paths at a party, a Cape Town t...  
2      After crossing paths at a party, a Cape Town t...  
3      After crossing paths at a party, a Cape Town t...  
4      After crossing paths at a party, a Cape Town t...  
...                                                  ...  
89267  A scrappy but poor boy worms his way into a ty...  
89268  A scrappy but poor boy worms his way into a ty...  
89269  A scrappy but poor boy worms his way into a ty...  
89270  A scrappy but poor boy worms his way into a ty...  
89271  A scrappy but poor boy worms his way into a ty...  

[89272 rows x 12 columns]

#### Column country has been unnested

#### d) Unnesting listed_in

In [36]:
df = df.assign(listed_in = df['listed_in'].str.split(', ')).explode('listed_in')

In [37]:
df = df.reset_index()

In [38]:
df.drop('index', axis=1, inplace=True)

In [39]:
df

show_id     type                 title         director  \
0           s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1           s2  TV Show         Blood & Water              NaN   
2           s2  TV Show         Blood & Water              NaN   
3           s2  TV Show         Blood & Water              NaN   
4           s2  TV Show         Blood & Water              NaN   
...        ...      ...                   ...              ...   
201758   s8807    Movie                Zubaan      Mozez Singh   
201759   s8807    Movie                Zubaan      Mozez Singh   
201760   s8807    Movie                Zubaan      Mozez Singh   
201761   s8807    Movie                Zubaan      Mozez Singh   
201762   s8807    Movie                Zubaan      Mozez Singh   

                         cast        country          date_added  \
0                         NaN  United States  September 25, 2021   
1                  Ama Qamata   South Africa  September 24, 2021   
2                  Ama Qamata   South Africa  September 24, 2021   
3                  Ama Qamata   South Africa  September 24, 2021   
4                 Khosi Ngema   South Africa  September 24, 2021   
...                       ...            ...                 ...   
201758         Anita Shabdish          India       March 2, 2019   
201759         Anita Shabdish          India       March 2, 2019   
201760  Chittaranjan Tripathy          India       March 2, 2019   
201761  Chittaranjan Tripathy          India       March 2, 2019   
201762  Chittaranjan Tripathy          India       March 2, 2019   

        release_year rating   duration               listed_in  \
0               2020  PG-13     90 min           Documentaries   
1               2021  TV-MA  2 Seasons  International TV Shows   
2               2021  TV-MA  2 Seasons               TV Dramas   
3               2021  TV-MA  2 Seasons            TV Mysteries   
4               2021  TV-MA  2 Seasons  International TV Shows   
...              ...    ...        ...                     ...   
201758          2015  TV-14    111 min    International Movies   
201759          2015  TV-14    111 min        Music & Musicals   
201760          2015  TV-14    111 min                  Dramas   
201761          2015  TV-14    111 min    International Movies   
201762          2015  TV-14    111 min        Music & Musicals   

                                              description  
0       As her father nears the end of his life, filmm...  
1       After crossing paths at a party, a Cape Town t...  
2       After crossing paths at a party, a Cape Town t...  
3       After crossing paths at a party, a Cape Town t...  
4       After crossing paths at a party, a Cape Town t...  
...                                                   ...  
201758  A scrappy but poor boy worms his way into a ty...  
201759  A scrappy but poor boy worms his way into a ty...  
201760  A scrappy but poor boy worms his way into a ty...  
201761  A scrappy but poor boy worms his way into a ty...  
201762  A scrappy but poor boy worms his way into a ty...  

[201763 rows x 12 columns]

### Revisiting Missing values

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201763 entries, 0 to 201762
Data columns (total 12 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   show_id       201763 non-null  object
 1   type          201763 non-null  object
 2   title         201763 non-null  object
 3   director      151338 non-null  object
 4   cast          199617 non-null  object
 5   country       189869 non-null  object
 6   date_added    201763 non-null  object
 7   release_year  201763 non-null  int64 
 8   rating        201763 non-null  object
 9   duration      201763 non-null  object
 10  listed_in     201763 non-null  object
 11  description   201763 non-null  object
dtypes: int64(1), object(11)
memory usage: 18.5+ MB


In [41]:
df.isnull().sum()/len(df)*100

show_id          0.000000
type             0.000000
title            0.000000
director        24.992194
cast             1.063624
country          5.895035
date_added       0.000000
release_year     0.000000
rating           0.000000
duration         0.000000
listed_in        0.000000
description      0.000000
dtype: float64

In [42]:
na_variables = [ var for var in df.columns if df[var].isnull().mean() > 0 ]
na_variables

['director', 'cast', 'country']

In [43]:
df[na_variables].isnull().mean()*100

director    24.992194
cast         1.063624
country      5.895035
dtype: float64

In [44]:
df['director'].value_counts()

Martin Scorsese        419
Youssef Chahine        409
Cathy Garcia-Molina    356
Steven Spielberg       355
Lars von Trier         336
                      ... 
Jerry Kolber             1
Marty Callner            1
Helena Coan              1
Will Allen               1
Fabio Ock                1
Name: director, Length: 4991, dtype: int64

In [45]:
df['director'].mode()

0    Martin Scorsese
dtype: object

In [46]:
df.iloc[np.where((pd.isna(df['country'])==True)  & (pd.isna(df['director'])==True))]

show_id     type                                title director  \
85          s4  TV Show                Jailbirds New Orleans      NaN   
86          s4  TV Show                Jailbirds New Orleans      NaN   
353        s11  TV Show  Vendetta: Truth, Lies and The Mafia      NaN   
354        s11  TV Show  Vendetta: Truth, Lies and The Mafia      NaN   
355        s11  TV Show  Vendetta: Truth, Lies and The Mafia      NaN   
...        ...      ...                                  ...      ...   
201196   s8786  TV Show                                  YOM      NaN   
201197   s8786  TV Show                                  YOM      NaN   
201704   s8804  TV Show                          Zombie Dumb      NaN   
201705   s8804  TV Show                          Zombie Dumb      NaN   
201706   s8804  TV Show                          Zombie Dumb      NaN   

              cast country          date_added  release_year rating  \
85             NaN     NaN  September 24, 2021          2021  TV-MA   
86             NaN     NaN  September 24, 2021          2021  TV-MA   
353            NaN     NaN  September 24, 2021          2021  TV-MA   
354            NaN     NaN  September 24, 2021          2021  TV-MA   
355            NaN     NaN  September 24, 2021          2021  TV-MA   
...            ...     ...                 ...           ...    ...   
201196  Mayur Vyas     NaN        June 7, 2018          2016  TV-Y7   
201197  Ketan Kava     NaN        June 7, 2018          2016  TV-Y7   
201704         NaN     NaN        July 1, 2019          2018  TV-Y7   
201705         NaN     NaN        July 1, 2019          2018  TV-Y7   
201706         NaN     NaN        July 1, 2019          2018  TV-Y7   

         duration               listed_in  \
85       1 Season              Docuseries   
86       1 Season              Reality TV   
353      1 Season          Crime TV Shows   
354      1 Season              Docuseries   
355      1 Season  International TV Shows   
...           ...                     ...   
201196   1 Season                Kids' TV   
201197   1 Season                Kids' TV   
201704  2 Seasons                Kids' TV   
201705  2 Seasons         Korean TV Shows   
201706  2 Seasons             TV Comedies   

                                              description  
85      Feuds, flirtations and toilet talk go down amo...  
86      Feuds, flirtations and toilet talk go down amo...  
353     Sicily boasts a bold "Anti-Mafia" coalition. B...  
354     Sicily boasts a bold "Anti-Mafia" coalition. B...  
355     Sicily boasts a bold "Anti-Mafia" coalition. B...  
...                                                   ...  
201196  With the mind of a human being, and the body o...  
201197  With the mind of a human being, and the body o...  
201704  While living alone in a spooky town, a young g...  
201705  While living alone in a spooky town, a young g...  
201706  While living alone in a spooky town, a young g...  

[4924 rows x 12 columns]

In [47]:
len(df)

201763

In [48]:
df[df['listed_in']=='Docuseries']

show_id     type                                            title  \
85          s4  TV Show                            Jailbirds New Orleans   
354        s11  TV Show              Vendetta: Truth, Lies and The Mafia   
497        s15  TV Show                  Crime Stories: India Detectives   
589        s21  TV Show  Monsters Inside: The 24 Faces of Billy Milligan   
653        s26  TV Show                             Love on the Spectrum   
...        ...      ...                                              ...   
200167   s8742  TV Show                                      Wild Arabia   
200475   s8756  TV Show                                Women Behind Bars   
200505   s8759  TV Show                           World's Busiest Cities   
200507   s8759  TV Show                           World's Busiest Cities   
200509   s8759  TV Show                           World's Busiest Cities   

               director              cast         country          date_added  \
85                  NaN               NaN             NaN  September 24, 2021   
354                 NaN               NaN             NaN  September 24, 2021   
497                 NaN               NaN             NaN  September 22, 2021   
589     Olivier Megaton               NaN             NaN  September 22, 2021   
653                 NaN  Brooke Satchwell       Australia  September 21, 2021   
...                 ...               ...             ...                 ...   
200167              NaN  Alexander Siddig  United Kingdom      March 31, 2017   
200475              NaN               NaN   United States    November 1, 2016   
200505              NaN        Anita Rani  United Kingdom    February 1, 2019   
200507              NaN      Ade Adepitan  United Kingdom    February 1, 2019   
200509              NaN          Dan Snow  United Kingdom    February 1, 2019   

        release_year rating   duration   listed_in  \
85              2021  TV-MA   1 Season  Docuseries   
354             2021  TV-MA   1 Season  Docuseries   
497             2021  TV-MA   1 Season  Docuseries   
589             2021  TV-14   1 Season  Docuseries   
653             2021  TV-14  2 Seasons  Docuseries   
...              ...    ...        ...         ...   
200167          2013  TV-PG   1 Season  Docuseries   
200475          2010  TV-14  3 Seasons  Docuseries   
200505          2017  TV-PG   1 Season  Docuseries   
200507          2017  TV-PG   1 Season  Docuseries   
200509          2017  TV-PG   1 Season  Docuseries   

                                              description  
85      Feuds, flirtations and toilet talk go down amo...  
354     Sicily boasts a bold "Anti-Mafia" coalition. B...  
497     Cameras following Bengaluru police on the job ...  
589     In the late 1970s, an accused serial rapist cl...  
653     Finding love can be hard for anyone. For young...  
...                                                   ...  
200167  The widely varied geology and dramatic landsca...  
200475  This reality series recounts true stories of w...  
200505  From Moscow to Mexico City, three BBC journali...  
200507  From Moscow to Mexico City, three BBC journali...  
200509  From Moscow to Mexico City, three BBC journali...  

[844 rows x 12 columns]

In [49]:
df.iloc[np.where((pd.isna(df['director'])==True) & (pd.isna(df['country'])==True))].shape[0]/len(df)*100

2.440487106159207

In [68]:
df.iloc[np.where((pd.isna(df['director'])==True) & (pd.isna(df['country'])==True) & (pd.isna(df['cast'])==False))]['director'].fillna(df.groupby(['cast'])['director'].agg(pd.Series.mode))

564      NaN
565      NaN
566      NaN
567      NaN
568      NaN
          ..
201193   NaN
201194   NaN
201195   NaN
201196   NaN
201197   NaN
Name: director, Length: 4740, dtype: float64

In [70]:
df.iloc[np.where((pd.isna(df['director'])==True) & (pd.isna(df['country'])==True) & (pd.isna(df['cast'])==False))]

show_id     type   title director               cast country  \
564        s20  TV Show  Jaguar      NaN      Blanca Suárez     NaN   
565        s20  TV Show  Jaguar      NaN      Blanca Suárez     NaN   
566        s20  TV Show  Jaguar      NaN      Blanca Suárez     NaN   
567        s20  TV Show  Jaguar      NaN        Iván Marcos     NaN   
568        s20  TV Show  Jaguar      NaN        Iván Marcos     NaN   
...        ...      ...     ...      ...                ...     ...   
201193   s8786  TV Show     YOM      NaN             Sairaj     NaN   
201194   s8786  TV Show     YOM      NaN  Devyani Dagaonkar     NaN   
201195   s8786  TV Show     YOM      NaN        Ketan Singh     NaN   
201196   s8786  TV Show     YOM      NaN         Mayur Vyas     NaN   
201197   s8786  TV Show     YOM      NaN         Ketan Kava     NaN   

                date_added  release_year rating  duration  \
564     September 22, 2021          2021  TV-MA  1 Season   
565     September 22, 2021          2021  TV-MA  1 Season   
566     September 22, 2021          2021  TV-MA  1 Season   
567     September 22, 2021          2021  TV-MA  1 Season   
568     September 22, 2021          2021  TV-MA  1 Season   
...                    ...           ...    ...       ...   
201193        June 7, 2018          2016  TV-Y7  1 Season   
201194        June 7, 2018          2016  TV-Y7  1 Season   
201195        June 7, 2018          2016  TV-Y7  1 Season   
201196        June 7, 2018          2016  TV-Y7  1 Season   
201197        June 7, 2018          2016  TV-Y7  1 Season   

                        listed_in  \
564        International TV Shows   
565     Spanish-Language TV Shows   
566         TV Action & Adventure   
567        International TV Shows   
568     Spanish-Language TV Shows   
...                           ...   
201193                   Kids' TV   
201194                   Kids' TV   
201195                   Kids' TV   
201196                   Kids' TV   
201197                   Kids' TV   

                                              description  
564     In the 1960s, a Holocaust survivor joins a gro...  
565     In the 1960s, a Holocaust survivor joins a gro...  
566     In the 1960s, a Holocaust survivor joins a gro...  
567     In the 1960s, a Holocaust survivor joins a gro...  
568     In the 1960s, a Holocaust survivor joins a gro...  
...                                                   ...  
201193  With the mind of a human being, and the body o...  
201194  With the mind of a human being, and the body o...  
201195  With the mind of a human being, and the body o...  
201196  With the mind of a human being, and the body o...  
201197  With the mind of a human being, and the body o...  

[4740 rows x 12 columns]

In [71]:
df.groupby(['cast'])['director'].agg(pd.Series.mode)

cast
 Jr.                          Sam Macaroni
"Riley" Lakdhar Dridi    Rebecca Zlotowski
'Najite Dede                  Aniedi Anwah
2 Chainz                                []
2Mex                          Ava DuVernay
                               ...        
Şevket Çoruh                  Ozan Açıktan
Şinasi Yurtsever            Selçuk Aydemir
Şükran Ovalı                Yılmaz Erdoğan
Şükrü Özyıldız              Yılmaz Erdoğan
Ṣọpẹ́ Dìrísù                   Remi Weekes
Name: director, Length: 36392, dtype: object

In [ ]:
df.groupby(['country'])['director'].agg(pd.Series.mode)

In [52]:
df[df['cast']=='Mayur Vyas']

show_id     type                           title     director  \
23318     s941    Movie  Motu Patlu: Deep Sea Adventure  Suhas Kadav   
23319     s941    Movie  Motu Patlu: Deep Sea Adventure  Suhas Kadav   
24860    s1007    Movie        Motu Patlu Dino Invasion  Suhas Kadav   
24861    s1007    Movie        Motu Patlu Dino Invasion  Suhas Kadav   
24868    s1008    Movie     Motu Patlu in Octupus World  Suhas Kadav   
24869    s1008    Movie     Motu Patlu in Octupus World  Suhas Kadav   
26732    s1073    Movie  Motu Patlu in the City of Gold  Suhas Kadav   
26733    s1073    Movie  Motu Patlu in the City of Gold  Suhas Kadav   
26734    s1073    Movie  Motu Patlu in the City of Gold  Suhas Kadav   
201196   s8786  TV Show                             YOM          NaN   

              cast country      date_added  release_year rating  duration  \
23318   Mayur Vyas     NaN     May 1, 2021          2014  TV-Y7    76 min   
23319   Mayur Vyas     NaN     May 1, 2021          2014  TV-Y7    76 min   
24860   Mayur Vyas     NaN  April 20, 2021          2018  TV-Y7    80 min   
24861   Mayur Vyas     NaN  April 20, 2021          2018  TV-Y7    80 min   
24868   Mayur Vyas     NaN  April 20, 2021          2017   TV-Y    81 min   
24869   Mayur Vyas     NaN  April 20, 2021          2017   TV-Y    81 min   
26732   Mayur Vyas     NaN  April 13, 2021          2018  TV-Y7    77 min   
26733   Mayur Vyas     NaN  April 13, 2021          2018  TV-Y7    77 min   
26734   Mayur Vyas     NaN  April 13, 2021          2018  TV-Y7    77 min   
201196  Mayur Vyas     NaN    June 7, 2018          2016  TV-Y7  1 Season   

                       listed_in  \
23318   Children & Family Movies   
23319                   Comedies   
24860   Children & Family Movies   
24861                   Comedies   
24868   Children & Family Movies   
24869                   Comedies   
26732   Children & Family Movies   
26733                   Comedies   
26734           Music & Musicals   
201196                  Kids' TV   

                                              description  
23318   Friends Motu and Patlu get more maritime excit...  
23319   Friends Motu and Patlu get more maritime excit...  
24860   A time machine sends Motu and Patlu back to th...  
24861   A time machine sends Motu and Patlu back to th...  
24868   While returning a goldfish and an octopus from...  
24869   While returning a goldfish and an octopus from...  
26732   Defeated by the strong yet naive Motu in a rac...  
26733   Defeated by the strong yet naive Motu in a rac...  
26734   Defeated by the strong yet naive Motu in a rac...  
201196  With the mind of a human being, and the body o...